File: apps/windows/TeslaSync.App/shared-surfaces/SourceLayerBadge.Model.cs
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.DataDisplay;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// One immutable badge sample — the two inputs the web <c>&lt;SourceLayerBadge&gt;</c> renders from
/// (web/src/components/data-display/SourceLayerBadge.tsx): the layered live-state <c>source</c> string
/// (<c>l1</c>/<c>l2</c>/<c>log</c>/<c>stale</c>, or null/unrecognized which projects to the unknown badge) and the
/// optional <c>ageMs</c> value age surfaced in the tooltip. It is the data the P1/S8
/// <see cref="ISourceLayerBadgeSource"/> exposes and the pure <see cref="SourceLayerBadgeProjection.Project"/>
/// consumes. <see cref="FromRepositoryResult{T}"/> derives the sample from a cache-then-network
/// <see cref="RepositoryResult{T}"/> by pulling the source layer (and optional age) out of the latest value,
/// exactly as a web diagnostics surface passes a signal's source metadata into the badge. Pure data — no WinUI
/// types — so it is unit-tested without a UI host.
/// </summary>
/// <param name="Source">The wire source-layer string (web <c>source</c>), or null for the unknown badge.</param>
/// <param name="AgeMs">The value age in milliseconds (web <c>ageMs</c>), or null when no age is known.</param>
public sealed record SourceLayerBadgeSnapshot(string? Source, double? AgeMs)
{
    /// <summary>The no-value sample — a null source, which projects to the unknown em-dash badge (web <c>source == null</c>).</summary>
    public static SourceLayerBadgeSnapshot Empty { get; } = new((string?)null, null);

    /// <summary>Create a sample for a known source layer with no age (the common debugger call site).</summary>
    /// <param name="source">The wire source-layer string (web <c>source</c>).</param>
    public static SourceLayerBadgeSnapshot Of(string? source) => new(source, null);

    /// <summary>Create a sample for a known source layer with a value age in milliseconds.</summary>
    /// <param name="source">The wire source-layer string (web <c>source</c>).</param>
    /// <param name="ageMs">The value age in milliseconds (web <c>ageMs</c>).</param>
    public static SourceLayerBadgeSnapshot Of(string? source, double ageMs) => new(source, ageMs);

    /// <summary>
    /// Derive a sample from a cache-then-network <see cref="RepositoryResult{T}"/> by selecting the source layer
    /// (and optional value age) out of the latest value — the native wiring for "where did this signal value come
    /// from". When the result carries a value (cached, refreshing, loaded or offline-cached) the selectors pull
    /// the source and age out of it; when there is no value yet (initial load, a success-but-empty response, or a
    /// hard failure with no cache) the sample is the no-value <see cref="Empty"/> equivalent, which projects to
    /// the unknown em-dash badge (the always-visible default — never a hidden surface, matching the web
    /// <c>source ?? 'unknown'</c> fallback). The whole derivation is WinUI-free so it is unit-tested against an
    /// in-memory result.
    /// </summary>
    /// <typeparam name="T">The repository's domain read-model type whose value carries the source metadata.</typeparam>
    /// <param name="result">The repository emission to read the latest value from.</param>
    /// <param name="selectSource">Selector that pulls the wire source-layer string out of the value.</param>
    /// <param name="selectAgeMs">Optional selector that pulls the value age (ms) out of the value; null when unused.</param>
    public static SourceLayerBadgeSnapshot FromRepositoryResult<T>(
        RepositoryResult<T> result,
        Func<T, string?> selectSource,
        Func<T, double?>? selectAgeMs = null)
    {
        ArgumentNullException.ThrowIfNull(result);
        ArgumentNullException.ThrowIfNull(selectSource);

        // A source layer exists only in the cache-then-network value-bearing states; Loading / Empty / Error
        // carry none and fall back to the unknown badge (HasValue keys off Value != null, which is unreliable for
        // a value-type T whose default is non-null).
        var hasValue = result.Status is LoadStatus.Cached or LoadStatus.Refreshing or LoadStatus.Loaded or LoadStatus.Offline;
        if (!hasValue)
        {
            return Empty;
        }

        var value = result.Value!;
        return new SourceLayerBadgeSnapshot(selectSource(value), selectAgeMs?.Invoke(value));
    }
}

/// <summary>
/// Canonical metadata for the SourceLayerBadge surface — the native analogue of the module-level <c>STYLE</c>
/// table, the <c>min-w</c> widths and the <c>formatAge</c> tiers in
/// web/src/components/data-display/SourceLayerBadge.tsx. Carries the diagnostics slug, the automation id, the
/// per-layer description i18n keys (each with the English fallback the web source renders verbatim — shared with
/// the <c>Strings/{lang}/Resources.resw</c> <c>translation.sourceLayer.*</c> catalog), the age-word key, the
/// lowercase layer tokens (the web <c>data-source</c> attribute), and the badge metrics. The per-layer glyph and
/// the token brush key are sourced from the shared <see cref="SourceLayers"/> logic so the surface, the atomic
/// <c>TsSourceLayerBadge</c> and the debugger tables all tint and label identically. UI-free so it is asserted in
/// tests.
/// </summary>
public static class SourceLayerBadgeRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "SourceLayerBadge";

    /// <summary>The root automation id Narrator and UI-automation resolve the surface by (web <c>data-testid</c>).</summary>
    public const string RootAutomationId = "source-layer-badge";

    /// <summary>ARIA role the surface exposes — a read-only diagnostics status indicator (native idiom; the web span carries no role).</summary>
    public const string StatusRole = "status";

    /// <summary>i18n key for the "age" tooltip word (web <c>t('sourceLayer.age', 'age')</c>).</summary>
    public const string AgeKey = "translation.sourceLayer.age";

    /// <summary>English fallback for <see cref="AgeKey"/> — the web literal.</summary>
    public const string AgeFallback = "age";

    /// <summary>Minimum badge width when the label is spelled out (web <c>min-w-[2.5rem]</c>, 40 DIPs).</summary>
    public const double ExpandedMinWidth = 40;

    /// <summary>Minimum badge width for the compact glyph (web <c>min-w-[1.5rem]</c>, 24 DIPs).</summary>
    public const double CompactMinWidth = 24;

    /// <summary>Badge glyph/label font size (web <c>text-[10px]</c>).</summary>
    public const double FontSize = 10;

    /// <summary>Corner radius fallback for the badge (web <c>rounded</c>, 4 DIPs) when the token is absent.</summary>
    public const double CornerRadiusFallback = 4;

    /// <summary>Letter spacing in 1/1000 em (web <c>tracking-wider</c> ≈ 0.05em).</summary>
    public const int CharacterSpacing = 50;

    /// <summary>The lowercase layer token the badge reports for query-by-attribute (web <c>data-source</c>).</summary>
    public static string LayerToken(SourceLayer layer) => layer switch
    {
        SourceLayer.L1 => "l1",
        SourceLayer.L2 => "l2",
        SourceLayer.Log => "log",
        SourceLayer.Stale => "stale",
        _ => "unknown",
    };

    /// <summary>i18n key for a layer's tooltip description (web <c>STYLE[key].descKey</c>).</summary>
    public static string DescriptionKey(SourceLayer layer) => layer switch
    {
        SourceLayer.L1 => "translation.sourceLayer.l1.desc",
        SourceLayer.L2 => "translation.sourceLayer.l2.desc",
        SourceLayer.Log => "translation.sourceLayer.log.desc",
        SourceLayer.Stale => "translation.sourceLayer.stale.desc",
        _ => "translation.sourceLayer.unknown.desc",
    };

    /// <summary>English fallback for a layer's tooltip description (web <c>STYLE[key].descFallback</c>; the shared <see cref="SourceLayers"/> description).</summary>
    public static string DescriptionFallback(SourceLayer layer) => SourceLayers.Tokens(layer).Description;

    /// <summary>The compact glyph/label a layer renders (web <c>STYLE[key].label</c>): L1 / L2 / LOG / STALE / em dash.</summary>
    public static string Label(SourceLayer layer) => SourceLayers.Tokens(layer).Label;

    /// <summary>The generated design-token brush key a layer tints from (web <c>STYLE[key].tint</c>).</summary>
    public static string AccentBrushKey(SourceLayer layer) => SourceLayers.Tokens(layer).AccentBrushKey;
}

/// <summary>
/// The fully projected, render-ready view of a <see cref="SourceLayerBadgeSnapshot"/> — everything the web
/// component derives before returning JSX (web/src/components/data-display/SourceLayerBadge.tsx): the resolved
/// <see cref="Layer"/> and its lowercase <see cref="SourceToken"/> (web <c>data-source</c>), the compact
/// <see cref="Label"/> glyph (web <c>style.label</c>), the token <see cref="AccentBrushKey"/> the badge tints
/// from, whether the label is spelled out (<see cref="ShowLabel"/>) and the resulting <see cref="MinWidth"/>
/// (web <c>min-w</c>), the localized relative <see cref="AgeText"/> (web <c>formatAge(ageMs)</c>), the composed
/// hover/Narrator <see cref="Tooltip"/> (web <c>tooltip</c>) and the accessible <see cref="AutomationName"/>.
/// Pure value type so every field is asserted headlessly.
/// </summary>
public readonly record struct SourceLayerBadgeProjection
{
    private SourceLayerBadgeProjection(
        SourceLayer layer,
        string sourceToken,
        string label,
        string accentBrushKey,
        bool showLabel,
        double minWidth,
        string? ageText,
        string description,
        string tooltip,
        string automationName)
    {
        Layer = layer;
        SourceToken = sourceToken;
        Label = label;
        AccentBrushKey = accentBrushKey;
        ShowLabel = showLabel;
        MinWidth = minWidth;
        AgeText = ageText;
        Description = description;
        Tooltip = tooltip;
        AutomationName = automationName;
    }

    /// <summary>The resolved source layer (web <c>STYLE[key] ?? unknown</c>).</summary>
    public SourceLayer Layer { get; }

    /// <summary>The lowercase source token for query-by-attribute (web <c>data-source={key}</c>).</summary>
    public string SourceToken { get; }

    /// <summary>The compact glyph/label (web <c>style.label</c>): L1 / L2 / LOG / STALE / em dash.</summary>
    public string Label { get; }

    /// <summary>The generated design-token brush key the badge tints from (web <c>style.tint</c>).</summary>
    public string AccentBrushKey { get; }

    /// <summary>Whether the label is spelled out — drives the wider min-width (web <c>showLabel</c>).</summary>
    public bool ShowLabel { get; }

    /// <summary>The badge minimum width in DIPs (web <c>min-w-[2.5rem]</c> / <c>min-w-[1.5rem]</c>).</summary>
    public double MinWidth { get; }

    /// <summary>The localized relative value age (web <c>formatAge(ageMs)</c>), or null when no age is known.</summary>
    public string? AgeText { get; }

    /// <summary>The localized layer description (web <c>t(style.descKey, style.descFallback)</c>).</summary>
    public string Description { get; }

    /// <summary>The composed hover / Narrator tooltip (web <c>tooltip</c>): description, plus the age clause when present.</summary>
    public string Tooltip { get; }

    /// <summary>The accessible name the automation peer reports (the composed tooltip — the colour-only badge's semantic).</summary>
    public string AutomationName { get; }

    /// <summary>
    /// Project a sample into a render-ready value, reproducing the web component body exactly
    /// (web/src/components/data-display/SourceLayerBadge.tsx L75-100): the <c>(source ?? 'unknown').toLowerCase()</c>
    /// key, the <c>STYLE[key] ?? unknown</c> classification (via the shared <see cref="SourceLayers.Parse"/>), the
    /// <c>formatAge(ageMs)</c> relative age (via the shared <see cref="FreshnessLogic.FormatSourceAge"/>), the
    /// description + age tooltip composition and the <c>min-w</c> width.
    /// </summary>
    /// <param name="snapshot">The badge sample (web <c>source</c> + <c>ageMs</c> props).</param>
    /// <param name="showLabel">Whether the label is spelled out (web <c>showLabel</c>).</param>
    /// <param name="localizer">The i18n facade every string resolves through.</param>
    public static SourceLayerBadgeProjection Project(
        SourceLayerBadgeSnapshot snapshot,
        bool showLabel,
        ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(snapshot);
        ArgumentNullException.ThrowIfNull(localizer);

        var layer = SourceLayers.Parse(snapshot.Source);

        // web: key = (source ?? 'unknown').toLowerCase(); the data-source attribute is the lowercased raw input.
        var sourceToken = (snapshot.Source ?? "unknown").ToLowerInvariant();

        var description = localizer.GetString(
            SourceLayerBadgeRegistration.DescriptionKey(layer),
            SourceLayerBadgeRegistration.DescriptionFallback(layer));

        var ageText = FreshnessLogic.FormatSourceAge(snapshot.AgeMs);

        // web: ageText ? `${desc} (${t('sourceLayer.age','age')}: ${ageText})` : desc.
        string tooltip;
        if (ageText is null)
        {
            tooltip = description;
        }
        else
        {
            var ageWord = localizer.GetString(SourceLayerBadgeRegistration.AgeKey, SourceLayerBadgeRegistration.AgeFallback);
            tooltip = $"{description} ({ageWord}: {ageText})";
        }

        var minWidth = showLabel
            ? SourceLayerBadgeRegistration.ExpandedMinWidth
            : SourceLayerBadgeRegistration.CompactMinWidth;

        return new SourceLayerBadgeProjection(
            layer: layer,
            sourceToken: sourceToken,
            label: SourceLayerBadgeRegistration.Label(layer),
            accentBrushKey: SourceLayerBadgeRegistration.AccentBrushKey(layer),
            showLabel: showLabel,
            minWidth: minWidth,
            ageText: ageText,
            description: description,
            tooltip: tooltip,
            automationName: tooltip);
    }
}

/// <summary>
/// PII-safe diagnostics for the SourceLayerBadge surface (P1/S11 diagnostics contract). The badge carries no user
/// content (only a source-layer token and a relative age), so the collector records ONLY the operational
/// <c>view.opened</c> event with the surface slug — never the source value or the age. Thread-safe; mirrors the
/// peer surfaces' diagnostics collectors.
/// </summary>
public sealed class SourceLayerBadgeDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public SourceLayerBadgeDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=SourceLayerBadge</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={SourceLayerBadgeRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/SourceLayerBadge/SourceLayerBadgeModel.kt
// Pure, framework-free model + projection + diagnostics for the SourceLayerBadge shared surface — the native
// analogue of every decision the web component makes (web/src/components/data-display/SourceLayerBadge.tsx)
// before it paints. No Compose, no Android framework, no platform clock, no HTTP: every declaration here is
// exercised off-device in the :app:testReleaseUnitTest gate, keeping the composable a thin render layer over
// these pure functions (the accepted sibling-surface contract — see FreshnessIndicator / Delta).
//
// What the web source is (and therefore the COMPLETE branch set this surface reproduces):
//   • A PURE PRESENTATIONAL primitive: a debugger-only badge showing where a live signal value came from. The
//     parent owns the `source` layer string + an optional `ageMs` and passes them in as props; the only hook
//     is `useTranslation` (i18n), so there is NO data port to bind — no P1/S8 state holder, no Source/
//     ViewModel. Modelling a fetch would invent behaviour the web spec does not have (honesty covenant: no
//     scope narrowing, no silent drift). The sibling presentational ports FreshnessIndicator / BatteryDelta /
//     AnimatedNumber document the same rationale (composable + model, no Source).
//   • Five source variants, each a fixed glyph + tint: l1 → "L1" (success/green), l2 → "L2" (info/blue),
//     log → "LOG" (muted), stale → "STALE" (warning/amber), and an unknown fallback → "—" (muted) for a null
//     / unrecognised layer ([SignalSourceLayer] + [sourceLayerGlyph] + [sourceLayerTint]). The lowercase
//     normalisation tolerates any backend casing exactly as the web `(source ?? 'unknown').toLowerCase()` does.
//   • `formatAge(ms)` → null when the age is absent, else a compact read-out: "{ms} ms" (< 1 s),
//     "{s} s" (< 1 min), "{min} min" (< 1 h), "{h} h" (< 1 d), "{d} d" — reproduced verbatim by
//     [formatSourceAge], including the web `Math.round` for the minute bucket and the `toFixed(1)` one-decimal
//     forms. The web `toFixed` is locale-independent (always a "." separator), so the decimal forms are
//     rendered with [Locale.ROOT] for byte-for-byte parity; the "ms"/"s"/"min"/"h"/"d" suffixes are unit
//     symbols the web hardcodes with no `t()` call (the prompt extracts ZERO i18n keys for them), so they are
//     symbol constants here, exactly like the unknown em-dash glyph — not localized microcopy.
//   • Tooltip / accessible description: the localized layer description, with " ({age}: {ageText})" appended
//     when an age is present (web `${desc} (${t('sourceLayer.age')}: ${ageText})`). The render layer resolves
//     the description + the "age" label from the shared P1/S10 catalog
//     (translation_sourceLayer_{l1,l2,log,stale,unknown}_desc + translation_sourceLayer_age); the model carries
//     only the layer + the formatted age so it stays framework-free and fully unit-tested.
//
// The surface's reproduced STATES are the five source-layer branches the web renders: l1 (the live, freshest
// state), l2 and log (cross-pod / durable-history reads), stale (the genuine amber "older than the freshness
// window" branch the prompt calls out), and unknown — the null / unrecognised-layer branch that doubles as
// the empty AND pre-data branch (a parent that has not yet resolved a layer passes null, exactly as the web
// `source ?? 'unknown'` path renders the muted "—" badge). The generic loading / error / offline lifecycle is
// intentionally absent: this surface fetches nothing, so a spinner or a retry affordance would fabricate
// behaviour the web spec does not have. Each branch is reduced here and asserted off-device, which therefore
// doubles as the per-state snapshot.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/SourceLayerBadge — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen segment and a PascalCase leaf are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName`
// is suppressed for the co-located supporting declarations alongside the namesake projection.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.sourcelayerbadge

import io.teslasync.shared.core.diagnostics.Logger
import java.util.Locale
import kotlin.math.roundToLong

/** The glyph shown for the unknown layer (web `STYLE.unknown.label = '—'`). A typographic symbol, not a word. */
const val SOURCE_LAYER_UNKNOWN_GLYPH: String = "\u2014"

/**
 * The layer a live signal value was satisfied from — the native mirror of the web `SignalSource`
 * (`'l1' | 'l2' | 'log' | 'stale' | string`). The render boundary maps this onto the glyph, the tint, and the
 * localized description. Any unrecognised / null backend value collapses to [Unknown] (web
 * `STYLE[key] ?? STYLE.unknown`).
 */
enum class SignalSourceLayer {
    /** Web `l1` — read from the in-process SignalStore (hot path, freshest). Green/success tint. */
    L1,

    /** Web `l2` — read from the Redis cross-pod cache (legacy entry, freshness unknown). Blue/info tint. */
    L2,

    /** Web `log` — replayed from signal_log (durable history). Muted tint. */
    Log,

    /** Web `stale` — Redis-backed value older than the 2-minute freshness window. Amber/warning tint. */
    Stale,

    /** Web fallback — null / unrecognised layer. Muted tint, em-dash glyph. */
    Unknown,
}

/**
 * Semantic tint role for a [SignalSourceLayer] — kept framework-free (no Compose `Color`) so the off-device
 * gate verifies the layer → tint mapping without a Compose host. The render boundary resolves each role to a
 * per-theme colour drawn from the TeslaTokens status palette / the Material scheme, so light / dark /
 * high-contrast all stay correct (mirrors the web background-15, text-200, border-30 alpha tints).
 */
enum class SourceLayerTint {
    /** Green — the freshest L1 read (web emerald). */
    Success,

    /** Blue — the L2 cross-pod cache read (web blue). */
    Info,

    /** Amber — the stale read past the freshness window (web amber). */
    Warning,

    /** Muted on-surface-variant — the log replay and the unknown fallback (web `--surface-2`/`--text-secondary`). */
    Muted,
}

/**
 * The fully reduced, render-ready projection of the badge — everything the composable needs, derived purely so
 * every branch is covered off-device. The view only resolves the tint colour, the localized description, and
 * lays out the glyph.
 *
 * @property layer the parsed source layer (web `key`), driving the glyph + tint + which description to resolve.
 * @property glyph the short uppercase chip label (web `style.label`): "L1" / "L2" / "LOG" / "STALE" / "—".
 * @property tint the semantic tint role the render layer maps onto a per-theme colour.
 * @property ageText the compact formatted age appended to the tooltip, or `null` when no age was supplied.
 */
data class SourceLayerProjection(
    val layer: SignalSourceLayer,
    val glyph: String,
    val tint: SourceLayerTint,
    val ageText: String?,
)

private const val MILLIS_PER_SECOND: Double = 1_000.0
private const val MILLIS_PER_MINUTE: Double = 60_000.0
private const val MILLIS_PER_HOUR: Double = 3_600_000.0
private const val MILLIS_PER_DAY: Double = 86_400_000.0

/**
 * Parses the backend source-layer string onto [SignalSourceLayer] — the native mirror of the web
 * `(source ?? 'unknown').toLowerCase()` lookup. Trims + lowercases so any backend casing/whitespace resolves,
 * and tolerates unknown / null values by falling back to [SignalSourceLayer.Unknown] (web
 * `STYLE[key] ?? STYLE.unknown`).
 */
fun parseSourceLayer(source: String?): SignalSourceLayer =
    when (source?.trim()?.lowercase(Locale.ROOT)) {
        "l1" -> SignalSourceLayer.L1
        "l2" -> SignalSourceLayer.L2
        "log" -> SignalSourceLayer.Log
        "stale" -> SignalSourceLayer.Stale
        else -> SignalSourceLayer.Unknown
    }

/** The short uppercase chip glyph for a [layer] — web `style.label` ("L1" / "L2" / "LOG" / "STALE" / "—"). */
fun sourceLayerGlyph(layer: SignalSourceLayer): String =
    when (layer) {
        SignalSourceLayer.L1 -> "L1"
        SignalSourceLayer.L2 -> "L2"
        SignalSourceLayer.Log -> "LOG"
        SignalSourceLayer.Stale -> "STALE"
        SignalSourceLayer.Unknown -> SOURCE_LAYER_UNKNOWN_GLYPH
    }

/**
 * The semantic tint role for a [layer] — the native mirror of the web per-variant tint (l1 → emerald,
 * l2 → blue, stale → amber, log + unknown → muted). The render boundary resolves the role to a per-theme
 * colour so the badge stays legible in every theme.
 */
fun sourceLayerTint(layer: SignalSourceLayer): SourceLayerTint =
    when (layer) {
        SignalSourceLayer.L1 -> SourceLayerTint.Success
        SignalSourceLayer.L2 -> SourceLayerTint.Info
        SignalSourceLayer.Stale -> SourceLayerTint.Warning
        SignalSourceLayer.Log -> SourceLayerTint.Muted
        SignalSourceLayer.Unknown -> SourceLayerTint.Muted
    }

/**
 * Human-readable age for a value in milliseconds — the native mirror of the web `formatAge`. Returns `null`
 * when [ageMs] is absent (web `ms == null || !Number.isFinite(ms)`; a `Long` is always finite, so only the
 * null case applies). Otherwise: "{ms} ms" (< 1 s), "{s} s" (< 1 min, one decimal), "{min} min" (< 1 h,
 * rounded), "{h} h" (< 1 d, one decimal), "{d} d" (one decimal). The minute bucket uses round-half-up
 * ([roundToLong] ties toward +∞, matching the web `Math.round`); the decimal forms format with [Locale.ROOT]
 * so the "." separator matches the web `toFixed(1)` byte-for-byte regardless of the device locale (the web
 * `toFixed` is locale-independent, so there is deliberately no locale knob to drift the separator).
 */
fun formatSourceAge(ageMs: Long?): String? {
    if (ageMs == null) return null
    return when {
        ageMs < MILLIS_PER_SECOND -> "$ageMs ms"
        ageMs < MILLIS_PER_MINUTE -> "${oneDecimal(ageMs / MILLIS_PER_SECOND)} s"
        ageMs < MILLIS_PER_HOUR -> "${(ageMs / MILLIS_PER_MINUTE).roundToLong()} min"
        ageMs < MILLIS_PER_DAY -> "${oneDecimal(ageMs / MILLIS_PER_HOUR)} h"
        else -> "${oneDecimal(ageMs / MILLIS_PER_DAY)} d"
    }
}

/** One-decimal fixed formatting — the native mirror of the web `Number.toFixed(1)` (locale-stable "."). */
private fun oneDecimal(value: Double): String = String.format(Locale.ROOT, "%.1f", value)

/**
 * Reduce a [source] layer string + an optional [ageMs] into the render-ready [SourceLayerProjection]. Pure
 * (no Compose), so every branch — l1 / l2 / log / stale / unknown, with and without an age — is covered by the
 * off-device gate and the composable stays a thin render layer.
 */
fun projectSourceLayerBadge(
    source: String?,
    ageMs: Long?,
): SourceLayerProjection {
    val layer = parseSourceLayer(source)
    return SourceLayerProjection(
        layer = layer,
        glyph = sourceLayerGlyph(layer),
        tint = sourceLayerTint(layer),
        ageText = formatSourceAge(ageMs),
    )
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the source
 * layer or the age — so a diagnostics line can never leak which signal a power user was inspecting or how
 * fresh it was. Kept free of Compose so it is unit-tested with a recording [Logger]; the composable calls it
 * once per surface open.
 */
object SourceLayerBadgeDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event — the slug the prompt mandates. */
    const val SLUG: String = "SourceLayerBadge"

    private const val VIEW_OPENED: String = "view.opened"
    private const val SURFACE_KEY: String = "surface"

    /** Emits the one PII-safe `view.opened` diagnostic. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
